2-repositories/iglu-server/src/main/scala/com/snowplowanalytics/iglu/server/storage/Storage.scala

```diff
@@ -0,0 +1,102 @@
+
+import com.snowplowanalytics.iglu.core.SchemaMap
+import com.snowplowanalytics.iglu.server.Config.StorageConfig
+import com.snowplowanalytics.iglu.server.model.{Permission, Schema, SchemaDraft}
+import com.snowplowanalytics.iglu.server.model.SchemaDraft.DraftId
+
+import scala.concurrent.ExecutionContext
+
+trait Storage[F[_]] {
+
+  def getSchema(schemaMap: SchemaMap)(implicit F: Monad[F]): F[Option[Schema]]
+  def getSchemasByVendor(vendor: String, wildcard: Boolean)(implicit F: Monad[F]): Stream[F, Schema] =
+    if (wildcard) getSchemas.filter(_.schemaMap.schemaKey.vendor.startsWith(vendor))
+    else getSchemas.filter(_.schemaMap.schemaKey.vendor === vendor )
+  def getSchemasByVendorName(vendor: String, name: String)(implicit F: Monad[F]): Stream[F, Schema] =
+    getSchemasByVendor(vendor, false).filter(_.schemaMap.schemaKey.name === name)
+  def getSchemas(implicit F: Monad[F]): Stream[F, Schema]
+  def getSchemaBody(schemaMap: SchemaMap)(implicit F: Monad[F]): F[Option[Json]] =
+    getSchema(schemaMap).nested.map(_.body).value
+  def addSchema(schemaMap: SchemaMap, body: Json, isPublic: Boolean)(implicit C: Clock[F], M: Monad[F]): F[Unit]
+
+  def addDraft(draftId: DraftId, body: Json, isPublic: Boolean)(implicit C: Clock[F], M: Monad[F]): F[Unit]
+  def getDraft(draftId: DraftId)(implicit F: Monad[F]): F[Option[SchemaDraft]]
+  def getDrafts(implicit F: Monad[F]): Stream[F, SchemaDraft]
+
+  def getPermission(apiKey: UUID)(implicit F: Monad[F]): F[Option[Permission]]
+  def addPermission(uuid: UUID, permission: Permission)(implicit F: Monad[F]): F[Unit]
+  def addKeyPair(keyPair: Permission.KeyPair, vendor: Permission.Vendor)(implicit F: Monad[F]): F[Unit] =
+    for {
+      _ <- addPermission(keyPair.read, Permission.ReadOnlyAny.copy(vendor = vendor))
+      _ <- addPermission(keyPair.write, Permission.Write.copy(vendor = vendor))
+    } yield ()
+  def deletePermission(uuid: UUID)(implicit F: Monad[F]): F[Unit]
+}
+
+object Storage {
+
+  /** Storage returned an object that cannot be parsed */
+  case class IncompatibleStorage(message: String) extends Throwable
+
+  /**
+    *
+    * @param transactEC a second ExecutionContext for executing JDBC operations
+    * @param config
+    * @tparam F
+    * @return
+    */
+  def initialize[F[_]: Effect: ContextShift](transactEC: ExecutionContext)
+                                            (config: StorageConfig): Resource[F, Storage[F]] = {
+    config match {
+      case StorageConfig.Dummy =>
+        Resource.liftF(storage.InMemory.empty)
+      case StorageConfig.Postgres(host, port, name, username, password, driver, threads, maxPoolSize) =>
+        for {
+          connectEC <- ExecutionContexts.fixedThreadPool(threads.getOrElse(32))
+          _ <- Resource.liftF(Async[F].delay(Class.forName(driver)))
+          xa <- HikariTransactor.initial[F](connectEC, transactEC)
+          _ <- Resource.liftF {
+            xa.configure { ds =>
+              Async[F].delay {
```

Comment:
`Async` is a super-strange choice here. It doesn't conform "least poweful abstraction" concept because `Async` is more poweful than `Sync` - first abstraction that has `delay`. It doesn't conform "conciseness" principle because we already have `Effect`, which implies `delay` as well.